include/LightGBM/utils/common.h

```diff
@@ -330,6 +331,26 @@ inline static const char* Atof(const char* p, double* out) {
   return p;
 }
 
+// Use fast_double_parse and strtod (if parse failed) to parse double.
+inline static const char* AtofPrecise(const char* p, double* out) {
```

Comment:
Does that mean we can replace
https://github.com/microsoft/LightGBM/blob/1d2f3e11018dcba56a8670f07c5d91f642e0266e/include/LightGBM/utils/common.h#L1079-L1102
which was added in #3942 with a call to this `AtofPrecise`? Do they behave exactly the same?

Reply:
Here a test program https://gist.github.com/cyfdecyf/63f4e7339bbe5a5a23474fda66375742

`strtod` returns signaling nan for "nan" and the like. For other case the behavior is the same.

I'll update `AtofPrecise` to return quiet nan.

Refer to: [1](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n975.htm#7.20.1.3%20the%20strtod,%20strtof,%20and%20strtold%20functions) [2](http://port70.net/~nsz/c/c11/n1570.html#7.22.1.3)

Reply:
I'm not an expert on floating number. So I write a test program https://gist.github.com/cyfdecyf/63f4e7339bbe5a5a23474fda66375742

The only difference is that `strtod` would not return negative NaN. So the `Atof` function in this gist handles this special case.

Please help take a look at the gist and check if there's any problem. I'll update `Common::Atof` and replace the change added in #3492.

Refer to: [1](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n975.htm#7.20.1.3%20the%20strtod,%20strtof,%20and%20strtold%20functions) [2](http://port70.net/~nsz/c/c11/n1570.html#7.22.1.3)

Reply:
The special handling for -NaN in gist (revision 5) the has one problem though: it's incorrect to handle input like " -nan" (note beginning space). But I'm wonder if there's need for the special handling of this case? From this mailling list [thread](https://newlib.sourceware.narkive.com/CpbSqKGq/strtod-nan-returns-negative-nan), it seems like different C library have different treatment for parsing "-nan".

I suggest just leave `Common::AtofPrecise` as is and don't add special handling for "-nan" and the like.

Reply:
Code introduced in #3942 is replaced by `AtofPrecise` in commit 498090dd.

Reply:
Great! LGTM.